lib/_stream_writable.js

```diff
@@ -700,6 +700,32 @@ function onCorkedFinish(corkReq, state, err) {
   state.corkedRequestsFree.next = corkReq;
 }
 
+function onFinished(stream, state, cb) {
+  if (state.destroyed && state.errorEmitted) {
+    // TODO(ronag): Backwards compat. Should be moved to end() without
+    // errorEmitted check and with errorOrDestroy.
+    const err = new ERR_STREAM_DESTROYED('end');
+    process.nextTick(cb, err);
+    return;
+  }
+
+  function onerror(err) {
+    stream.removeListener('finish', onfinish);
+    stream.removeListener('error', onerror);
+    cb(err);
+    if (stream.listenerCount('error') === 0) {
+      stream.emit('error', err);
+    }
+  }
+  function onfinish() {
+    stream.removeListener('finish', onfinish);
+    stream.removeListener('error', onerror);
+    cb();
+  }
+  stream.on('finish', onfinish);
+  stream.on('error', onerror);
+}
```

Comment:
There is a side effect of listening for `'error'` here. If somebody is doing `once('error', fn)` they are going to see that error "bubble up" to `'uncaughtException'`. I'm not sure if it's something we can handle easily, however we should document this behavior in the docs.

Reply:
Good catch. I'll give it some thought.

Reply:
@mcollina: It might be possible the do this properly but it would require an array of callbacks on the state object. Is it worth it?

Reply:
Another alternative is to not allow end to be called multiple times and on the first end call queue an "empty" write and it's callback.

Reply:
Actually, I think it might be possible to avoid this edge case by using `prependListener`?

Reply:
> Actually, I think it might be possible to avoid this edge case by using prependListener('error', onerror)?

I'll be ok with that. Would you mind adding a unit test for that case?

Reply:
Fixed